### PR TITLE
EF-24: Owned entity collections

### PR DESCRIPTION
**Pull request by damieng**
Implementation of owned entity collections.

The main issue here was handling the synthetic non-stored keys on the collection which as well as having the parent key like non-collections also needed another value. We use a generator even though we don't persist or read these values to keep EF happy.

### CODE REVIEW

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Query/OwnedEntityTests.cs [hunk 1]
```diff
@@ -96,12 +96,14 @@ public void OwnedEntity_nested_one_level_creates()
         }
     }
 
-    [Fact(Skip = "We do not yet support updating owned types")]
-    public void OwnedEntity_nested_collection_creates()
+    [Fact]
```
Comment (sanych-sun) [inline tests/MongoDB.EntityFrameworkCore.FunctionalTests/Query/OwnedEntityTests.cs:99]: We probably need another test for the case when an entity has 2 navigation properties of same type. Like: 
```
class Person
{
  public Location PrimaryLocation { get; set; }
  public Location SecondaryLocation { get; set; }
}
```

File: tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/MongoValueGeneratorSelectorTests.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.ValueGeneration;
+using MongoDB.Bson;
+using MongoDB.EntityFrameworkCore.ValueGeneration;
+
+namespace MongoDB.EntityFrameworkCore.UnitTests.ValueGeneration;
+
+public class MongoValueGeneratorSelectorTests
+{
+    private readonly IModel __model;
```
Comment (sanych-sun) [inline tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/MongoValueGeneratorSelectorTests.cs:26]: Just curious what double-underscores means here?

File: tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/OwnedEntityIndexValueGeneratorTests.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+using MongoDB.EntityFrameworkCore.ValueGeneration;
+
+namespace MongoDB.EntityFrameworkCore.UnitTests.ValueGeneration;
+
+public class OwnedEntityIndexValueGeneratorTests
+{
+    [Fact]
+    public void Next_generates_non_temporary_id()
```
Comment (sanych-sun) [inline tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/OwnedEntityIndexValueGeneratorTests.cs:23]: Should we add tests to ensure no duplicates generated for concurrent execution?

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoEntityTypeBuilderExtensions.cs [hunk 4]
```diff
@@ -115,4 +116,77 @@ public static class MongoEntityTypeBuilderExtensions
...
+    /// <returns>The same builder instance if the configuration was applied, <see langword="null" /> otherwise.</returns>
+    public static IConventionEntityTypeBuilder? SetElementName(
+        this IConventionEntityTypeBuilder entityTypeBuilder,
+        string? name,
+        bool fromDataAnnotation = false)
+    {
+        if (!entityTypeBuilder.CanSetElementName(name, fromDataAnnotation))
+        {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoEntityTypeBuilderExtensions.cs:167]: It does seem odd but that's what the Cosmos one does https://github.com/dotnet/efcore/blob/release/8.0/src/EFCore.Cosmos/Extensions/CosmosPropertyBuilderExtensions.cs#L80 - frankly I'd have expected it to throw...

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Query/OwnedEntityTests.cs (hunk 1 above)
Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.FunctionalTests/Query/OwnedEntityTests.cs:99]: Can't add this one yet as it would fail - it's a known bug EF-58.

File: tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/MongoValueGeneratorSelectorTests.cs (hunk 2 above)
Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/MongoValueGeneratorSelectorTests.cs:26]: Double-underscore is private static which comes from the naming rules in the .editorconfig we borrowed from the Mongo C# Driver repository. Here however their name should have been corrected when they stopped being static, will fix.

File: tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/OwnedEntityIndexValueGeneratorTests.cs (hunk 3 above)
Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.UnitTests/ValueGeneration/OwnedEntityIndexValueGeneratorTests.cs:23]: We could but we don't currently have any of that level of test coverage and this interlock pattern is what they use throughout EF Core.

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoEntityTypeBuilderExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoEntityTypeBuilderExtensions.cs:167]: `IConvention*Builder` types are not Fluent API and returning `null` is expected in cases where an overriding configuration is present.

[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoEntityTypeBuilderExtensions.cs:167]: It might be, but from user prospective: this is extension method to IConventionPropertyBuilder which returns IConventionPropertyBuilder... it looks like I can chain more configuration code (that's why I see this as fluent). And if I chain more code, it might result to NullReferenceException, instead of some more reasonable exception. So we might want to either to change it to be void or throw there for consistency.

[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoEntityTypeBuilderExtensions.cs:167]: Whenever one of these methods returns `null` further configuration should be avoided, so the calls should be chained using `?.` instead of `.`
